### ISSUE TITLE
async-std@1.6.0-beta.1: panic waiting for TcpStream::connect

### ISSUE THREAD

**Issue by jimblandy** [async-rs/async-std#785]
Steps to reproduce:

```
$ cat Cargo.toml
[package]
name = "async-std-bug"
version = "0.1.0"
authors = ["Jim Blandy <jimb@red-bean.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
async-std = "1.6.0-beta.1"
$ cat src/main.rs
use async_std::{io, net, task};

fn main() -> io::Result<()> {
    task::block_on(async {
        let _socket = net::TcpStream::connect("localhost:8080").await?;
        Ok(())
    })
}
$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/async-std-bug`
thread 'main' panicked at 'cannot convert a blocking task into `JoinHandle`', <::std::macros::panic macros>:2:4
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
$ RUST_BACKTRACE=full cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/async-std-bug`
thread 'main' panicked at 'cannot convert a blocking task into `JoinHandle`', <::std::macros::panic macros>:2:4
stack backtrace:
   0:     0x55c1db449e64 - backtrace::backtrace::libunwind::trace::h5d52ba5f20882f09
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
   1:     0x55c1db449e64 - backtrace::backtrace::trace_unsynchronized::hceee092869668a74
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
   2:     0x55c1db449e64 - std::sys_common::backtrace::_print_fmt::ha312c2904605e4d5
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x55c1db449e64 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h5b9981092140b727
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x55c1db464d2c - core::fmt::write::h5f6d7d8de88b4173
                               at src/libcore/fmt/mod.rs:1063
   5:     0x55c1db4441d3 - std::io::Write::write_fmt::h893169117de3cc15
                               at src/libstd/io/mod.rs:1426
   6:     0x55c1db44c655 - std::sys_common::backtrace::_print::h8ab61d4120f7a335
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x55c1db44c655 - std::sys_common::backtrace::print::h8aae19fbb153bf2a
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x55c1db44c655 - std::panicking::default_hook::{{closure}}::h1ee5b7d8b6f83429
                               at src/libstd/panicking.rs:204
   9:     0x55c1db44c3a2 - std::panicking::default_hook::hd6c32c13403f9210
                               at src/libstd/panicking.rs:224
  10:     0x55c1db44ccb2 - std::panicking::rust_panic_with_hook::h1f2449d529a25f22
                               at src/libstd/panicking.rs:470
  11:     0x55c1db421d11 - std::panicking::begin_panic::h41cabb1350be9a1c
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/panicking.rs:397
  12:     0x55c1db390025 - <smol::task::Task<T> as core::convert::Into<async_task::join_handle::JoinHandle<T,()>>>::into::h6f8f4f40a1343c48
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/<::std::macros::panic macros>:2
  13:     0x55c1db398ee9 - async_std::task::spawn_blocking::spawn_blocking::h446d93064ceb5adf
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/spawn_blocking.rs:40
  14:     0x55c1db38cde7 - <str as async_std::net::addr::ToSocketAddrs>::to_socket_addrs::h9426f72a794192f4
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/net/addr.rs:237
  15:     0x55c1db370e11 - <&T as async_std::net::addr::ToSocketAddrs>::to_socket_addrs::hcbac3d73762bad59
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/net/addr.rs:267
  16:     0x55c1db3827bd - async_std::net::tcp::stream::TcpStream::connect::{{closure}}::hbcf5ba0d4b797436
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/net/tcp/stream.rs:75
  17:     0x55c1db37f83e - <std::future::GenFuture<T> as core::future::future::Future>::poll::hc1cfb9ad760cc4a1
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/future.rs:44
  18:     0x55c1db37f128 - std::future::poll_with_tls_context::hfbe91a656d28cfdc
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/future.rs:102
  19:     0x55c1db384362 - async_std_bug::main::{{closure}}::hb6c2906ee1c3cdf8
                               at src/main.rs:5
  20:     0x55c1db37f980 - <std::future::GenFuture<T> as core::future::future::Future>::poll::hef906f61de5c830c
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/future.rs:44
  21:     0x55c1db37a98e - <async_std::task::builder::SupportTaskLocals<F> as core::future::future::Future>::poll::{{closure}}::h0c599a8c31e11565
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/builder.rs:174
  22:     0x55c1db3724b7 - async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current::{{closure}}::hbca830e0c7dd52d1
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/task_locals_wrapper.rs:60
  23:     0x55c1db373736 - std::thread::local::LocalKey<T>::try_with::hd38e0f380e68d087
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/thread/local.rs:262
  24:     0x55c1db372802 - std::thread::local::LocalKey<T>::with::h787a06e0a4f35a47
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/thread/local.rs:239
  25:     0x55c1db37227f - async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current::h7354f739f1c69794
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/task_locals_wrapper.rs:55
  26:     0x55c1db37a92b - <async_std::task::builder::SupportTaskLocals<F> as core::future::future::Future>::poll::h9c4593c973107ffe
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/builder.rs:172
  27:     0x55c1db379ff6 - smol::run::run::{{closure}}::{{closure}}::hcc3488213c900411
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/run.rs:144
  28:     0x55c1db380677 - scoped_tls_hkt::ScopedKey<T>::set::h54a4b231333e3eb2
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-hkt-0.1.2/src/lib.rs:488
  29:     0x55c1db3831a0 - smol::throttle::setup::h542390c48b603ab1
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/throttle.rs:31
  30:     0x55c1db3798a6 - smol::run::run::{{closure}}::he2ea545dd92ec261
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/run.rs:144
  31:     0x55c1db37c1a1 - core::ops::function::FnOnce::call_once::h93e83bb06d3ee094
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libcore/ops/function.rs:232
  32:     0x55c1db383208 - smol::context::enter::h3e3361b612e6017b
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/context.rs:8
  33:     0x55c1db379f8f - smol::run::run::{{closure}}::{{closure}}::h401d3bb966c955f7
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/run.rs:112
  34:     0x55c1db38055f - scoped_tls_hkt::ScopedKey<T>::set::h48313bb3d0eca140
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-hkt-0.1.2/src/lib.rs:488
  35:     0x55c1db379255 - smol::thread_local::ThreadLocalExecutor::enter::hbdab804eed779fba
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/thread_local.rs:57
  36:     0x55c1db37973f - smol::run::run::{{closure}}::h312fea4ad5b1f027
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/run.rs:112
  37:     0x55c1db379f3f - smol::run::run::{{closure}}::{{closure}}::h34b90a8a84596d40
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/run.rs:113
  38:     0x55c1db374a4f - smol::work_stealing::WORKER::<impl smol::work_stealing::WORKER>::set::h619fc8b3b0cd0077
                               at /home/jimb/rust/book/non-meta/async/bugs/async-std-bug/<::scoped_tls_hkt::scoped_thread_local macros>:40
  39:     0x55c1db374bb5 - smol::work_stealing::Worker::enter::h3ad43516ad71e102
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/work_stealing.rs:149
  40:     0x55c1db3797af - smol::run::run::{{closure}}::hdd6b9d00b87b077c
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/run.rs:113
  41:     0x55c1db3795e4 - smol::run::run::h7a54e63f58bb966f
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.7/src/run.rs:115
  42:     0x55c1db37b02b - async_std::task::builder::Builder::blocking::{{closure}}::hb37e2087ff445e84
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/builder.rs:154
  43:     0x55c1db372383 - async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current::{{closure}}::h2941126daafd5944
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/task_locals_wrapper.rs:60
  44:     0x55c1db372d5f - std::thread::local::LocalKey<T>::try_with::h44f1bab794a9b21d
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/thread/local.rs:262
  45:     0x55c1db3725a2 - std::thread::local::LocalKey<T>::with::h220983d0567ce990
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/thread/local.rs:239
  46:     0x55c1db372227 - async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current::h1fcb8a21721da4b3
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/task_locals_wrapper.rs:55
  47:     0x55c1db37af7a - async_std::task::builder::Builder::blocking::h40e819abd28abdba
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/builder.rs:154
  48:     0x55c1db37b5a8 - async_std::task::block_on::block_on::h4e6de38dd0585ba5
                               at /home/jimb/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/block_on.rs:33
  49:     0x55c1db380243 - async_std_bug::main::h27622729810f54ac
                               at src/main.rs:4
  50:     0x55c1db37a07d - std::rt::lang_start::{{closure}}::hfc6a9da8652c7fb9
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/rt.rs:67
  51:     0x55c1db44c773 - std::rt::lang_start_internal::{{closure}}::h9a4aa16acf1cdc99
                               at src/libstd/rt.rs:52
  52:     0x55c1db44c773 - std::panicking::try::do_call::h0b6fc9f6090c1e2b
                               at src/libstd/panicking.rs:303
  53:     0x55c1db44fb87 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:86
  54:     0x55c1db44d1cc - std::panicking::try::h9eaeeaa81242ec77
                               at src/libstd/panicking.rs:281
  55:     0x55c1db44d1cc - std::panic::catch_unwind::h07d504c1b691e8fb
                               at src/libstd/panic.rs:394
  56:     0x55c1db44d1cc - std::rt::lang_start_internal::hcea4e704875ab132
                               at src/libstd/rt.rs:51
  57:     0x55c1db37a047 - std::rt::lang_start::h2b7469cb4afb485a
                               at /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/rt.rs:67
  58:     0x55c1db38027a - main
  59:     0x7f33531e4f43 - __libc_start_main
  60:     0x55c1db37073e - _start
  61:                0x0 - <unknown>
$ 
```

**Comment by jimblandy** [comment]
This is on Fedora.

**Comment by seirdb** [comment]
The problem seems to be that it’s to convert a `smol::Task::Inner::Blocking(..)` into a `async_task::JoinHandle`, which isn’t possible.

There seems to be 2 different kind of blocking tasks since the move to `smol`; the ones “native” to `smol`, and the implementation within `async-std` itself.

**Comment by seirdb** [comment]
The strange thing is that there’s a test on master (2b6c7fe) which is also broken if I run the tests on my setup (Ubuntu 19.10, rust 1.43.1).

```
---- fs::file::tests::async_file_drop stdout ----
thread 'fs::file::tests::async_file_drop' panicked at 'cannot convert a blocking task into `JoinHandle`', <::std::macros::panic macros>:2:4
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Yet on the CI it does seem to pass.

**Comment by seirdb** [comment]
This seems to be fixed with the latest release of `smol`, `0.1.8`.
